src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertLinqQueryToLinqMethodProvider<TQueryExpression> : CodeRefactoringProvider
+        where TQueryExpression : SyntaxNode
+    {
+        protected abstract string Title { get; }
+
+        protected abstract bool TryConvert(
+            TQueryExpression queryExpression,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken,
+            out DocumentUpdateInfo documentUpdate);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var queryExpression = root.FindNode(context.Span) as TQueryExpression;
```

Comment:
what happen if query is used as argument of a method call?

Reply:
just added this. please see in the next iteration :-)

---
In reply to: [176887084](https://github.com/dotnet/roslyn/pull/25362#discussion_r176887084) [](ancestors = 176887084)